internal/mirror/mirror.go

```diff
@@ -0,0 +1,276 @@
+
+			// go through again and process downloads.
+			for filename, status := range actions {
+				if deadlinePassed(deadline) {
+					logger.Warnw("mirror didn't complete processing before timeout", "mirrorID", mirror.ID)
+					break
+				}
+
+				if status.needsDownload() {
+					logger.Infow("mirroring export file", "mirrorID", mirror.ID, "file", filename)
+
+					client := &http.Client{
+						Timeout: s.config.ExportFileDownloadTimeout,
+					}
+					resp, err := client.Get(status.DownloadPath)
+					if err != nil {
+						logger.Errorw("unable to download file", "filename", filename, "error", err)
+						status.Failed = true
+						continue
+					}
+
+					defer resp.Body.Close()
+					bytes, err := ioutil.ReadAll(resp.Body)
+					if err != nil {
+						logger.Errorw("unable to read file", "filename", filename, "error", err)
+						status.Failed = true
+						continue
+					}
+
+					// Write this file to the blob store.
+					bsCtx, cancel := context.WithTimeout(ctx, time.Minute)
+					defer cancel()
+					objName := fmt.Sprintf("%s/%s", mirror.FilenameRoot, status.Filename)
+					if err := blobstore.CreateObject(bsCtx, mirror.CloudStorageBucket, objName, bytes, true, storage.ContentTypeZip); err != nil {
+						logger.Errorw("unable to write file to blobstore", "filename", filename, "error", err)
+						status.Failed = true
+						continue
+					}
+
+					status.Saved = true
+					indexObjects = append(indexObjects, status)
+				}
+			}
+
+			// Write new index file
+			sort.Slice(indexObjects, func(i, j int) bool {
+				return indexObjects[i].Order < indexObjects[j].Order
+			})
+			shortFilenames := make([]string, 0, len(indexObjects))
+			filenames := make([]string, 0, len(indexObjects))
+			for _, obj := range indexObjects {
+				// Only persist state of the files we got to.
+				if obj.Saved {
+					shortFilenames = append(shortFilenames, obj.Filename)
+					filenames = append(filenames, fmt.Sprintf("%s/%s", mirror.FilenameRoot, obj.Filename))
+				}
+			}
+			data := []byte(strings.Join(filenames, "\n"))
+			indexName := fmt.Sprintf("%s/index.txt", mirror.FilenameRoot)
+			bsCtx, cancel := context.WithTimeout(ctx, time.Minute)
+			defer cancel()
+			if err := blobstore.CreateObject(bsCtx, mirror.CloudStorageBucket, indexName, data, true, storage.ContentTypeTextPlain); err != nil {
+				logger.Errorw("unable to write index to blobstore", "filename", "index.txt", "error", err)
+				continue
+			}
+
+			// sync state to DB.
+			if err := s.mirrorDB.SaveFiles(ctx, mirror.ID, shortFilenames); err != nil {
+				logger.Errorw("error saving index state to database", "mirrorID", mirror.ID, "error", err)
+			}
+			logger.Infow("finished mirror", "mirrorID", mirror.ID, "index", mirror.IndexFile)
+		}
+
+		status := http.StatusOK
+		w.WriteHeader(status)
+		w.Write([]byte(http.StatusText(status)))
+	}
+}
+
+func deadlinePassed(deadline time.Time) bool {
+	return time.Now().After(deadline)
+}
```

Comment:
I'd really like to see this broken down into smaller functions that are easily tested. The current HTTP handler is very long and _very_ nested.

Reply:
can do as a follow up if that's OK

Reply:
yup